modules/storage/storage.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2022 Red Hat
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	corev1 "k8s.io/api/core/v1"
+)
+
+// ExtraVolumePropagation defines a const that represent the propagation policy
+type ExtraVolumePropagation string
+// ExtraVolType represents a "label" that can be optionally added to the VolMounts
+// instance
+type ExtraVolType string
+// ServiceType identifies the Service, Group or instance (e.g. the backend) that
+// receives an Extra Volume that can potentially be mounted
+type ServiceType string
+
+const (
+	// PropagationEverywhere is used to define a propagation policy that allows
+	// to get the volumes mounted to all the OpenStack services
+	PropagationEverywhere ExtraVolumePropagation = "All"
+	// DBSync represents a common ServiceType defined by the OpenStack operators
+	// that keeps track of the DBSync pod
+	DBSync                ServiceType            = "DBSync"
+	// Compute represents a common ServiceType that can be translated into an
+	// external-data-plane related propagation policy
+	Compute               ServiceType            = "Compute"
+)
+
+// VolMounts is the data structure used to expose Volumes and Mounts that can
+// be added to a pod according to the defined Propagation policy
+type VolMounts struct {
+	// +kubebuilder:validation:Optional
+	// Propagation defines which pod should mount the volume
+	Propagation  []ExtraVolumePropagation `json:"propagation,omitempty"`
+	// Label associated to a given extraMount
+	// +kubebuilder:validation:Optional
+	ExtraVolType ExtraVolType             `json:"extraVolType,omitempty"`
+	// +kubebuilder:validation:Required
+	Volumes      []corev1.Volume          `json:"volumes"`
+	// +kubebuilder:validation:Required
+	Mounts       []corev1.VolumeMount     `json:"mounts"`
+}
+
```

Comment:
Different k8s types with different calls made by operators (you can observe a pattern in all the operators, e.g.,  [1]), so ... I'd say we `could` have a unified function for both, but I'd like to coordinate this work (a follow up?) from the operators to lib-common, and not the reverse.
Even though I think you're point is a good one, I'd like to avoid a massive change in the operators, while having CI, API versioning, and things like that would help having smooth changes w/o introducing bugs.

[1] https://github.com/fmount/cinder-operator/blob/master/pkg/cinder/volumes.go

Reply:
Ok, I think I see your point, which is actually interesting: let me think a bit about it and I'll try to propose such a
valid update that cover this part.
Keep in mind that, in the current status (for the purpose of introducing the feature), I'm not looking into revisiting
the operators structure (e.g. have a single `GetVolume` to rule them all), so the code you see here is the result
of extending the operator existing functions and keep this logic into `lib-common`.
We can choose what's best to keep in `lib-common` and what is supposed to be operator related logic: the cleaner
we design this behavior, the better.
As per your comment here, we can have only `1 propagation function`, which actually returns a `VolMounts` array
that represents the list of `filtered VolMounts` (which contains `Volumes` and `VolumeMounts`): the resulting structure
can be merged w/ the current struct explicitly provided at the operator level.

Let me try an example to better clarify.
At `Glance` level (in `lib-common/modules/storage/glance.go`) I can have something like:

```
func (g *GlanceExtraVolMounts) Propagate(svc []PropagationType) []VolMounts {
...
...
}
```

which calls, behind the scenes, `storage.go` (that still represents the basic building block for Propagation), and,
according to the propagation list I can return a list of `VolMounts` that can be applied to the pods.

`storage.go` can be changed and  should have something like:

```
func (v *VolMounts) Propagate(svc []PropagationType) []VolMounts {

	var vl []VolMounts

	for _, p := range v.Propagation {
		if canPropagate(p, svc) {
			vl = append(vl, VolMounts {
				Volumes: v.Volumes,
				Mounts: v.Mounts,
			})
		}
	}

	return vl
}
```

and Glance can actually build/filter on top of the basic propagation function (e.g., processing additional parameters
in the future).
Similar thing should happen with Cinder, I used Glance in this example because it's easier.

The operator side of things (e.g., `pkg/glance/volumes.go`), however, should be something like:

```
        // for any GlanceExtraVolumes struct, propagate the VolMounts
	for _, exv := range extraVol {
		for _, vol := range exv.Propagate(svc) {
			vm = append(vm, vol.Mounts...)
		}
	}
```

`VolMounts` is an array, and the propagation functions are applied/called by the defined struct, hence an
iteration over there is necessary.
This approach simplify **a lot** the amount of code in the `lib-common` module, and still gives the opportunity
to build (in the future) on top of this feature.

Reply:
I feel like we are converging to the same idea.